### PR TITLE
adding duplicate removal and implimenting L1 quality cuts

### PR DESCRIPTION
**Pull request by Sam-Harper**
#### PR description:

This PR adds the l1 muon quality selection and l1 muon duplicate removal 

#### PR validation:

On 10000 MB events rate goes from  87+/-16kHz to 24kHz +/-8kHz for tripple mu 7_5_5 which now matches the L1 reference (https://twiki.cern.ch/twiki/bin/view/CMS/PhaseIIL1TriggerMenuTools)

### CODE REVIEW

File: HLTrigger/HLTfilters/plugins/L1TTkMuonFilter.cc [hunk 1]
```diff
@@ -26,6 +27,31 @@
 // constructors and destructor
 //
 
+namespace {
+  bool passMuQual(const l1t::TkMuon& muon) {
+    if (muon.muonDetector() == 3) {
+      int quality = muon.quality();
+      if (quality == 11 || quality == 13 || quality == 14 || quality == 15) {
```
Comment (fwyzard) [inline HLTrigger/HLTfilters/plugins/L1TTkMuonFilter.cc:34]: are these values defined anywhere, *e.g.* in some L1T package ?

**Comment by Martin-Grunewald** [comment]
please test

File: HLTrigger/HLTfilters/plugins/L1TTkMuonFilter.cc [hunk 2]
```diff
@@ -26,6 +27,31 @@
...
+      }
+    } else {
+      return true;
+    }
+  }
+  bool isDupMuon(const l1t::TkMuonRef& muon, const std::vector<l1t::TkMuonRef>& existing) {
+    for (const auto& exist : existing) {
+      //it is our understanding that there is an exact eta phi match
+      //and we should not be concerned with numerical precision
+      if (reco::deltaR2(*muon, *exist) <= 0) {
```
Comment (fwyzard) [inline HLTrigger/HLTfilters/plugins/L1TTkMuonFilter.cc:47]: are the duplicate muons always from the same subsystem (barrel/overlap.forward) ?
or are the coordinates exactly the same also for muons from different subsystems ?

File: HLTrigger/HLTfilters/plugins/L1TTkMuonFilter.cc [hunk 3]
```diff
@@ -34,6 +60,8 @@ L1TTkMuonFilter::L1TTkMuonFilter(const edm::ParameterSet& iConfig)
   min_N_ = iConfig.getParameter<int>("MinN");
   min_Eta_ = iConfig.getParameter<double>("MinEta");
   max_Eta_ = iConfig.getParameter<double>("MaxEta");
+  applyQuality_ = iConfig.getParameter<bool>("applyQuality");
+  doDupRemoval_ = iConfig.getParameter<bool>("doDupRemoval");
```
Comment (fwyzard) [inline HLTrigger/HLTfilters/plugins/L1TTkMuonFilter.cc:64]: I'd prefer `applyQualityCuts` and `applyDuplicateRemoval` for consistency among the new parameter names

not a strong opinion, though

File: HLTrigger/HLTfilters/plugins/L1TTkMuonFilter.cc (hunk 2 above)
Comment (Sam-Harper) [reply HLTrigger/HLTfilters/plugins/L1TTkMuonFilter.cc:47]: I dont know.  I do know the cut listed on the twiki is deltaR>0

File: HLTrigger/HLTfilters/plugins/L1TTkMuonFilter.cc (hunk 3 above)
Comment (Sam-Harper) [reply HLTrigger/HLTfilters/plugins/L1TTkMuonFilter.cc:64]: I dont really mind. I'm happy to make that change.

File: HLTrigger/HLTfilters/plugins/L1TTkMuonFilter.cc (hunk 1 above)
Comment (Sam-Harper) [reply HLTrigger/HLTfilters/plugins/L1TTkMuonFilter.cc:34]: no idea. I freely admit these magic numbers are not great and should move. What we could do is make it configurable with a vector "allowed qualities". 

Basically I read the twiki and applied the cuts? That is extent of my understanding here :)

**Comment by silviodonato** [comment]
@cms-sw/hlt-l2 the DAS errors are not related to this PR (they have been 'fixed' by #33020 )

**Comment by Martin-Grunewald** [comment]
+1

**Comment by Sam-Harper** [comment]
Hi Martin, thanks for the speedy signing but I still need to address Andrea's comments which I will do shortly

**Comment by Martin-Grunewald** [comment]
-1
apologies, @Sam-Harper

File: HLTrigger/HLTfilters/plugins/L1TTkMuonFilter.cc (hunk 1 above)
Comment (fwyzard) [reply HLTrigger/HLTfilters/plugins/L1TTkMuonFilter.cc:34]: Indeed, I haven't been able to find any definition that one could reuse in the L1 software

**Comment by Sam-Harper** [comment]
please test

**Comment by Martin-Grunewald** [comment]
+1

**Comment by qliphy** [comment]
+1

**Pull request merged**